toolkit/offchain/src/csl.rs

```diff
@@ -278,22 +307,65 @@ impl TransactionBuilderExt for TransactionBuilder {
 		Ok(())
 	}
 
-	fn set_required_fields_and_build(
+	fn balance_update_and_build(
 		&mut self,
 		ctx: &TransactionContext,
 	) -> Result<Transaction, JsError> {
-		self.add_collateral_inputs(ctx)?;
-		self.calc_script_data_hash(&convert_cost_models(
-			&ctx.protocol_parameters.plutus_cost_models,
-		))?;
-		self.add_required_signer(&ctx.payment_key_hash);
-		self.add_inputs_from_and_change_with_collateral_return(
-			&ctx.payment_utxos.to_csl()?,
-			cardano_serialization_lib::CoinSelectionStrategyCIP2::LargestFirstMultiAsset,
-			&ChangeConfig::new(&key_hash_address(&ctx.payment_key_hash, ctx.network)),
-			&ctx.protocol_parameters.collateral_percentage.into(),
-		)?;
-		self.build_tx()
+		// Tries if the largest UTXO is enough to cover collateral, if not, adds more UTXOs
+		// starting from the largest remaining.
+		fn select_collateral_inputs(
+			builder: &TransactionBuilder,
+			ctx: &TransactionContext,
+		) -> Result<Vec<OgmiosUtxo>, JsError> {
+			let max_possible_collaterals: Vec<_> = {
+				let mut utxos = ctx.payment_utxos.clone();
+				utxos.sort_by(|a, b| b.value.lovelace.cmp(&a.value.lovelace));
+				let max_inputs = ctx.protocol_parameters.max_collateral_inputs;
+				utxos
+					.into_iter()
+					.take(max_inputs.try_into().expect("max_collateral_input fit in usize"))
+					.collect()
+			};
+
+			let mut selected = vec![];
+			for input in max_possible_collaterals {
+				selected.push(input);
+				let mut builder = builder.clone();
+				builder.add_collateral_inputs(ctx, &selected)?;
+				// Check if the used inputs are enough
+				let result = builder.add_inputs_from_and_change_with_collateral_return(
+					&ctx.payment_utxos.to_csl()?,
+					CoinSelectionStrategyCIP2::LargestFirstMultiAsset,
+					&ChangeConfig::new(&key_hash_address(&ctx.payment_key_hash(), ctx.network)),
+					&ctx.protocol_parameters.collateral_percentage.into(),
+				);
+				if result.is_ok() {
+					break;
```

Comment:
Couldn't we just return `builder.build_tx()` at this point and avoid having to half duplicate this logic in `try_balance`?

Reply:
I agree